src/Microsoft.TestPlatform.CrossPlatEngine/Client/CriteriaTransform.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client
+{
+    using System.Collections.Generic;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Host;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client;
+    using System.Linq;
+
+    internal static class CriteriaTransform
+    {
+
+        public static void UpdateDiscoveryCriteria(DiscoveryCriteria discoveryCriteria, ITestRuntimeProvider testRuntimeProvider)
+        {
+            var actualTestSources = testRuntimeProvider.GetTestSources(discoveryCriteria.Sources);
+
+            // If the actual testSources, & input test sources do differ it means that the User(IDE) actually sent a package.
+            // We are limiting that only one package can be sent per session, so picking the first entry in sources
+            if (discoveryCriteria.Sources.Except(actualTestSources).Any())
+            {
+                discoveryCriteria.Package = discoveryCriteria.Sources.FirstOrDefault();
+
+                // Allow TestRuntimeProvider to update source map, this is required for remote scenarios.
+                // If we run for specific tests, then we expect the test case object to contain correct source path for remote scenario as well
+                UpdateTestSources(actualTestSources, discoveryCriteria.AdapterSourceMap);
+            }
+        }
+
+        public static void UpdateTestRunCriteriaForSources(TestRunCriteria testRunCriteria, ITestRuntimeProvider testRuntimeProvider, ref List<string> testPackages)
```

Comment:
Return TestRunCriteria and provide a Package member in this.

Reply:
moved to extension methods for Discovery and RunCriteria